File: static-data/www/onboarding/onboarding.js
/*
    Onionr - Private P2P Communication

    Handles onboarding for Onionr

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>
*/

fetch('/getnewkeys', {
    headers: {
        "token": webpass
    }})
    .then((resp) => resp.text())
    .then(function(resp) {
        keys = resp.split('')
})

function getCheckValue(elName){
    return document.getElementsByName(elName)[0].checked
}

function sendConfig(configInfo){
    fetch('/setonboarding', {
        method: 'POST',
        headers: {
          "token": webpass,
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({configInfo})
        }).then(function(data) {
            window.location.href = window.location.origin + '/' + window.location.hash
        })
}


document.getElementById('onboardingForm').onsubmit = function(e){
    submitInfo = {}
    submitInfo.massSurveil = getCheckValue('state')
    submitInfo.stateTarget = getCheckValue('stateTarget')
    submitInfo.localThreat = getCheckValue('local')
    submitInfo.networkContrib = getCheckValue('networkContribution')
    submitInfo.plainContrib = getCheckValue('networkContributionPlain')
    submitInfo.donate = getCheckValue('donate')
    //submitInfo.deterministic = getCheckValue('useDeterministic')
    submitInfo.mail = getCheckValue('useMail')
    submitInfo.circles = getCheckValue('useCircles')
    submitInfo.useDark = getCheckValue('useDarkTheme')

    if (submitInfo.donate){
        openDonateModal(submitInfo)
        return false
    }

    sendConfig(submitInfo)

    e.preventDefault()
}

/* Fix label clicking since bulma is weird */

let labelClickFix = function(labels) {
    for (i = 0; i < labels.length; i++){
        labels[i].onclick = function(event){
            document.getElementsByName(event.target.getAttribute("for"))[0].checked ^= 1
        }
    }

}
let setupLabelFix = function(){
    var labels = document.getElementsByTagName('label')
    var icons = document.getElementsByTagName('i')
    labelClickFix(labels)
    labelClickFix(icons)
}
setupLabelFix()

File: src/config/onboarding.py
"""Onionr - Private P2P Communication.

Setup config from onboarding choices
"""
from pathlib import Path
from typing import Union

from filepaths import onboarding_mark_file
from onionrtypes import JSONSerializable
from onionrtypes import OnboardingConfig
import config
"""
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""


def _get_val_or_none(json: dict, key: str) -> Union[None, JSONSerializable]:
    try:
        return json['configInfo'][key]
    except KeyError:
        return None


def set_config_from_onboarding(config_settings: OnboardingConfig):
    get = _get_val_or_none

    config.reload()

    if get(config_settings, 'optimize'):
        config.set('ui.animated_background', False)
        config.set('general.insert_deniable_blocks', False)

    if get(config_settings, 'stateTarget') or not get(config_settings,
                                                      'networkContrib'):
        config.set('general.security_level', 1)

    if get(config_settings, 'localThreat'):
        config.set('general.security_level', 3)
        config.set('transports.lan', False)

    config.set('ui.theme', 'light')
    if get(config_settings, 'useDark'):
        config.set('ui.theme', 'dark')

    disabled = config.get('plugins.disabled', [])

    if not get(config_settings, 'circles') or \
            config.get('general.security_level') > 0:
        disabled.append('flow')

    if not get(config_settings, 'mail'):
        disabled.append('pms')

    config.set('plugins.disabled', disabled)

    config.set('general.store_plaintext_blocks',
               get(config_settings, 'plainContrib'))

    config.set('onboarding.done', True, savefile=True)


def set_onboarding_finished():
    """Create the onboarding completed setting file"""
    Path(onboarding_mark_file).touch()


def is_onboarding_finished() -> bool:
    return True
